File: src/app/player.py
from typing import TYPE_CHECKING
from datetime import datetime, timedelta

if TYPE_CHECKING:
    from .game import Game

class Player:
    def __init__(self, game: "Game", name: str, is_ai: bool):
        self.game: "Game" = game
        self.name: str = name
        self.is_ai: bool = is_ai
        self.width: float = 20
        self.height: float = 200
        self.x: float = 0
        self.y: float = (game.height - self.height) / 2
        self.time_move: datetime = datetime.now()
        self.go_up: bool = False
        self.go_down: bool = False
        self.ai_target_y = self.y + self.height / 2
        self.score: int = 0

    def move(self):
        if self.is_ai:
            self.play_ai()

        if self.go_up:
            self.y -= 5
        if self.go_down:
            self.y += 5
        self.y = min(max(self.y, 0), self.game.height - self.height)

    def play_ai(self):
        if self.time_move + timedelta(seconds=1) <= datetime.now():
            self.time_move = datetime.now()
            ball = self.game.ball
            if ball.speed_x != 0 and ball.speed_x * (self.x - self.game.width / 2) >= 0:
                ball_x = ball.x
                if self.x > self.game.width / 2:
                    ball_x += ball.size
                ball_y = ball.y + ball.size / 2
                self.ai_target_y = (self.x - ball_x) * ball.speed_y / ball.speed_x + ball_y

        center_y = self.y + self.height / 2
        self.go_up = center_y > self.ai_target_y + 30
        self.go_down = center_y < self.ai_target_y - 30

    def to_json(self):
        return {
            "name": self.name,
            "x": self.x,
            "y": self.y,
            "width": self.width,
            "height": self.height,
            "score": self.score
        }

File: backend/src/Player.ts
import { Game } from "./Game";

export default class Player {
  static readonly WIDTH = 20;
  static readonly HEIGHT = 200;

  readonly id: string;
  private readonly game: Game;
  readonly name: string;
  readonly isAi: boolean;
  private speed: number = 5;
  x: number = 0;
  y: number = 0;
  goUp: boolean = false;
  goDown: boolean = false;

  private aiLastCheck: number = Date.now();
  private aiTargetY: number = 0;

  constructor(game: Game, name: string, isAi: boolean) {
    this.id = crypto.randomUUID();
    this.game = game;
    this.name = name;
    this.isAi = isAi;
    this.aiTargetY = ((Math.random() - 0.5) * Player.HEIGHT) + Game.HEIGHT / 2;
  }

  get isAtLeft() {
    return this.x - Game.WIDTH / 2 < 0;
  }

  get left() {
    return this.x;
  }

  get right() {
    return this.x + Player.WIDTH;
  }

  get top() {
    return this.y;
  }

  get bottom() {
    return this.y + Player.HEIGHT;
  }

  get centerX() {
    return this.x + Player.WIDTH / 2;
  }

  get centerY() {
    return this.y + Player.HEIGHT / 2;
  }

  move() {
    if (this.isAi) this.playAi();

    if (this.goUp) this.y -= this.speed;
    if (this.goDown) this.y += this.speed;
    this.y = Math.min(Math.max(this.y, 0), Game.HEIGHT - Player.HEIGHT);
  }

  playAi() {
    const ball = this.game.ball;
    if (Date.now() - this.aiLastCheck >= 1_000) {
      this.aiLastCheck = Date.now();

      if (ball.speedX != 0 && ball.goToLeft == this.isAtLeft) {
        this.speed = 5;
        this.aiTargetY = ((this.isAtLeft
            ? this.right - ball.left
            : this.left - ball.right)
          * ball.speedY) / ball.speedX + ball.centerY;
        if (this.aiTargetY < 0)
          this.aiTargetY = (-this.aiTargetY) * 1.25;
        else if (this.aiTargetY > Game.HEIGHT)
          this.aiTargetY = Game.HEIGHT - ((this.aiTargetY - Game.HEIGHT) * 1.25);
        this.aiTargetY += (Math.random() - 0.5) * 50;
      } else {
        this.speed = 2;
        this.aiTargetY = ball.centerY;
      }
    }
    this.goUp = this.centerY > this.aiTargetY + (Player.HEIGHT / 4);
    this.goDown = this.centerY < this.aiTargetY - (Player.HEIGHT / 4);
    if (!this.goUp && !this.goDown && ball.goToLeft != this.isAtLeft) {
      this.speed = 1;
      this.aiTargetY = Math.floor(Date.now() / 500) % 2 === 0 ? 0 : Game.HEIGHT;
    }
  }

  toJSON() {
    return {
      name: this.name,
      x: this.x,
      y: this.y,
      width: Player.WIDTH,
      height: Player.HEIGHT
    };
  }
}
